### PR TITLE
Fix filtering rounding error

### PR DESCRIPTION
**Pull request by DradeAW**
The `astype(dtype)` always converts int to the lower integer, which in some cases causes real problems.

This PR fixes that.

### CODE REVIEW

**Comment by DradeAW** [comment]
An example of how I found this error:

I was plotting the distribution of spikes amplitude, and wanted to compare it to the distribution of voltage on the same channel.

The bin 0 was twice bigger than what it should be, because of how `astype(np.int16)` works.
The bin x includes values between [x ; x+1). **Except** for the bin 0 which includes values between (-1; 1).

Rounding before converting to an int fixes this issue, and more globally makes things neater.

**Comment by zm711** [comment]
I take the point that you'd prefer this to behave like a normal round (I think that it is better too for my mental schema), but the standard for casting from float -> int (for example) is just to drop the decimal values. So you don't think that this might lead some users to be confused by the non-standard behavior of casting to an int, but seeing the value round? Just curious how you're thinking about this "round" occurring internal to the function without the user potentially knowing about it?

**Comment by DradeAW** [comment]
Well if I have a value at 12, and I apply a slight filter which reduces the value to 11.98, I expect the value to be 12 after casting to an integer ...

I don't think many users expect to receive 11?

**Comment by DradeAW** [comment]
The thing is that most recording are with int16 today. And by default in SI when you filter, you get back an int16.
And as I stated above, there is a problem with 0.

**Comment by zm711** [comment]
Actually I completely agree for Gaussian. In that case you use the dtype from the traces and so the user has no option. If the recording is an int16 and it gets filter the user is stuck with int16.

But in `filter` you are required to specify the dtype, so the user could choose to return the data with a dtype that maintains the filtered values. So my question is more just for filter where you could warn that user to return in float if they wanted to keep the decimal.

**Comment by DradeAW** [comment]
> But in `filter` you are required to specify the dtype

https://github.com/SpikeInterface/spikeinterface/blob/main/src/spikeinterface/preprocessing/filter.py#L70

I think a lot of users don't specify the `dtype` ^^
But let's see what @alejoe91 @samuelgarcia and @h-mayorquin think!

**Comment by samuelgarcia** [comment]
Hi.
thank you for this fix. It was a need since long time. I also in some cases enormous bin at zeros in distrib.
I think this is not only filtering but all preprocessing that preserve the input dtype.
Maybe we could make a larger check on this.

About a warning when do not use the decimal will be really painfull. For me this preserving the input dtype is a feature in spikeinterface.preprocessing since long time.

For me it has to be documented well.
We already have this in the doc : https://spikeinterface.readthedocs.io/en/latest/modules/preprocessing.html#impact-on-recording-dtype
For sure, Zach, will find a more adavnced formula for this

**Comment by DradeAW** [comment]
> I think this is not only filtering but all preprocessing that preserve the input dtype.
> Maybe we could make a larger check on this.

I can try to do that if you think that's best!

**Comment by zm711** [comment]
Okay fair enough. If it is occurring everywhere in the same manner and not ad hoc, then it makes sense.

>For me it has to be documented well.
We already have this in the doc : https://spikeinterface.readthedocs.io/en/latest/modules/preprocessing.html#impact-on-recording-dtype
For sure, Zach, will find a more adavnced formula for this

I can double check that part of the docs (I already found a typo reading it just now so I'll do another pass soon). And yeah I think just making it clear in the docs that this is what is happening is super important for users. I will say that I did add the `np.rint` in Neo for selecting frames for analog signals, so I'm definitely not opposed to this :) just curious about what is given as the responsibility/freedom of the user.

> About a warning when do not use the decimal will be really painfull. For me this preserving the input dtype is a feature in spikeinterface.preprocessing since long time.

Fair enough! If it is expected behavior for the library then people can't expect to be warned for the norm.

**Comment by h-mayorquin** [comment]
> The astype(dtype) always converts int to the lower integer, which in some cases causes real problems.

This is not true:
```python
# Random floats between -1 and 1
x = np.random.uniform(-2, 2, 10)
print(x)
x.astype("int")
output:
[-0.01363288  0.42181579  1.25074111  0.19889013  1.11397242 -0.95708403
  1.15950627 -1.3823785   1.43423315 -0.61537481]

```

I actually don't know enough about filtering to have an opinion on this. My own -weakly held- take now is that there should NOT be an option to force the output in a specific format:

https://github.com/SpikeInterface/spikeinterface/issues/914#issuecomment-1734761048

I think that the tricks that we use don't achieve saving memory:
https://github.com/SpikeInterface/spikeinterface/issues/2029
They are also confusing as as some point we built a test that was statistically wrong:
https://github.com/SpikeInterface/spikeinterface/issues/1972

And, the most subjective, tricks like the int scale in the Z-score hurt my mathematical sensibilities:
https://github.com/h-mayorquin/spikeinterface/blob/1ca0655e371a8dbc34c49ffde251b3628c03db8b/src/spikeinterface/preprocessing/normalize_scale.py#L239-L242

Looking forward to learn here, 
What is able to control the output meaning to achieve?

**Comment by DradeAW** [comment]
> This is not true:

```
>>> import numpy as np
>>> 
>>> np.random.uniform(-1, 1, 20).astype(np.int32)
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int32)
```

By lower I meant "closer to 0"

**Comment by h-mayorquin** [comment]
The correct claim is that casting behaves like floor for positive values and like a ceiling for negative values. There is not a conception of "lower" that I can map to that.

File: doc/modules/preprocessing.rst [hunk 1]
```diff
@@ -74,6 +74,8 @@ dtype (unless specified otherwise):
 
 Some scaling pre-processors, such as :code:`whiten()` or :code:`zscore()`, will force the output to :code:`float32`.
```
Comment (zm711) [inline doc/modules/preprocessing.rst:75]: This is part of your problem @h-mayorquin? The forcing of the output? Just so I can stay on the same page.

Comment (h-mayorquin) [reply doc/modules/preprocessing.rst:75]: Yes, I don't know why we have that option but my proble most likely is ignorance. Maybe there are good scenarios that I jus't don't understand.

Comment (zm711) [reply doc/modules/preprocessing.rst:75]: I honestly don't know either, so I'm happy to learn too :)

**Review by samuelgarcia — APPROVED** [review]

**Comment by samuelgarcia** [comment]
Merci beaucoup. This was important.

**Pull request merged**